### PR TITLE
Added support for trailing dot

### PR DESCRIPTION
**Pull request by gapra-msft**

### CODE REVIEW

File: cmd/removeEnumerator.go [hunk 1]
```diff
@@ -75,6 +71,9 @@ func newRemoveEnumerator(cca *CookedCopyCmdArgs) (enumerator *CopyEnumerator, er
 	if cca.IncludeAfter != nil {
 		filters = append(filters, &IncludeAfterDateFilter{Threshold: *cca.IncludeAfter})
 	}
+	if !cca.trailingDot && (cca.FromTo.To() == common.ELocation.File() || cca.FromTo.From() == common.ELocation.File()) {
+		filters = append(filters, &TrailingDotFilter{})
```
Comment (gapra-msft) [inline cmd/removeEnumerator.go:75]: open question: how to deal with files that have trailing dot if trailing dot is off

File: cmd/zc_traverser_file.go [hunk 2]
```diff
@@ -34,6 +34,8 @@ import (
 	"github.com/Azure/azure-storage-azcopy/v10/common"
 )
 
+const trailingDotErrMsg = "File share contains file/directory %s with a trailing dot but the trailing dot flag was not set to true. AzCopy will automatically filter this file/directory out to prevent data corruption."
```
Comment (adreed-msft) [inline cmd/zc_traverser_file.go:37]: Given we're now raw yes/no and not strip/include/filter, is this message accurate or necessary?

File: cmd/zc_traverser_file.go [hunk 3]
```diff
@@ -67,6 +70,10 @@ func (t *fileTraverser) Traverse(preprocessor objectMorpher, processor objectPro
 
 	// if not pointing to a share, check if we are pointing to a single file
 	if targetURLParts.DirectoryOrFilePath != "" {
+		if t.trailingDot != common.ETrailingDotOption.Enable() && strings.HasSuffix(targetURLParts.DirectoryOrFilePath, ".") {
```
Comment (adreed-msft) [inline cmd/zc_traverser_file.go:73]: Ditto

File: cmd/zc_traverser_file.go [hunk 4]
```diff
@@ -194,14 +201,24 @@ func (t *fileTraverser) Traverse(preprocessor objectMorpher, processor objectPro
 				return fmt.Errorf("cannot list files due to reason %s", err)
 			}
 			for _, fileInfo := range lResp.FileItems {
+				// These conditions are to prevent a GetProperties from returning a 404 when trailing dot is turned off.
```
Comment (adreed-msft) [inline cmd/zc_traverser_file.go:204]: Ditto

File: e2etest/factory.go [hunk 5]
```diff
@@ -60,8 +63,19 @@ func (TestResourceFactory) GetFileServiceURL(accountType AccountType) azfile.Ser
 	if err != nil {
 		panic(err)
 	}
-	pipeline := azfile.NewPipeline(credential, azfile.PipelineOptions{})
-	return azfile.NewServiceURL(*u, pipeline)
+	// Closest to API goes first; closest to the wire goes last
+	f := []pipeline.Factory{
```
Comment (adreed-msft) [inline e2etest/factory.go:67]: Can we not re-use the integrated new pipeline functions in AzCopy? Is there a good reason to duplicate this?

File: e2etest/helpers.go [hunk 6]
```diff
@@ -257,7 +258,18 @@ func newNullFolderCreationTracker() ste.FolderCreationTracker {
 func generateParentsForAzureFile(c asserter, fileURL azfile.FileURL) {
 	accountName, accountKey := GlobalInputManager{}.GetAccountAndKey(EAccountType.Standard())
 	credential, _ := azfile.NewSharedKeyCredential(accountName, accountKey)
-	err := ste.AzureFileParentDirCreator{}.CreateParentDirToRoot(ctx, fileURL, azfile.NewPipeline(credential, azfile.PipelineOptions{}), newNullFolderCreationTracker())
+	// Closest to API goes first; closest to the wire goes last
```
Comment (adreed-msft) [inline e2etest/helpers.go:261]: Ditto

File: e2etest/factory.go [hunk 7]
```diff
@@ -146,7 +160,18 @@ func (TestResourceFactory) GetFileShareULWithSAS(c asserter, accountType Account
 	fullURL, err := url.Parse(rawURL)
 	c.AssertNoErr(err)
 
-	return azfile.NewShareURL(*fullURL, azblob.NewPipeline(azblob.NewAnonymousCredential(), azblob.PipelineOptions{}))
+	// Closest to API goes first; closest to the wire goes last
```
Comment (adreed-msft) [inline e2etest/factory.go:163]: Ditto

File: e2etest/runner.go [hunk 8]
```diff
@@ -107,6 +107,9 @@ func (t *TestRunner) SetAllFlags(p params, o Operation) {
 	set("is-object-dir", p.isObjectDir, false)
 	set("debug-skip-files", strings.Join(p.debugSkipFiles, ";"), "")
 	set("check-md5", p.checkMd5.String(), "FailIfDifferent")
+	if p.trailingDot != nil {
```
Comment (adreed-msft) [inline e2etest/runner.go:110]: This pattern only really makes sense if the zero-value is not the default; does p.trailingDot need to be a pointer for some reason?

**Review by adreed-msft — APPROVED** [review]

File: e2etest/runner.go (hunk 8 above)
Comment (gapra-msft) [reply e2etest/runner.go:110]: This is because I need a way to test that if nothing is passed, the trailing dot files still work.

**Review by nakulkar-msft — APPROVED** [review]

File: cmd/zc_traverser_file.go (hunk 2 above)
Comment (gapra-msft) [reply cmd/zc_traverser_file.go:37]: You're right, let me fix the verbiage.

File: e2etest/factory.go (hunk 5 above)
Comment (gapra-msft) [reply e2etest/factory.go:67]: I can reuse the pipeline creation logic in ste

File: e2etest/runner.go (hunk 8 above)
Comment (adreed-msft) [reply e2etest/runner.go:110]: Right; but that's the default behavior. The set() function doesn't apply the parameter if it's default, iirc.

File: cmd/zc_traverser_file.go (hunk 3 above)
Comment (gapra-msft) [reply cmd/zc_traverser_file.go:73]: removed the nil but I will leave the log here as a warning

File: cmd/zc_traverser_file.go (hunk 4 above)
Comment (gapra-msft) [reply cmd/zc_traverser_file.go:204]: this is just a log for warning, so when customers experience weird behavior, they can look at the logs and see why

File: e2etest/runner.go (hunk 8 above)
Comment (gapra-msft) [reply e2etest/runner.go]: changed this to use set's behavior in the correct way

**Pull request merged**